include/deal.II/base/mpi_consensus_algorithms.templates.h

```diff
@@ -247,14 +233,88 @@ namespace Utilities
+        // it has found a message from any process with a given
+        // tag.
+        while (outstanding_answers.size() > 0)
+          {
+            const int tag_deliver = Utilities::MPI::internal::Tags::
+              consensus_algorithm_nbx_process_deliver;
+
+            int        request_is_pending;
+            MPI_Status status;
+            const auto ierr = MPI_Iprobe(MPI_ANY_SOURCE,
+                                         tag_deliver,
+                                         this->comm,
+                                         &request_is_pending,
+                                         &status);
+            AssertThrowMPI(ierr);
+
+            // If there is no pending message with this tag,
+            // then we are clearly not done receiving everything
+            // yet -- so return false.
+            if (request_is_pending == 0)
+              return false;
+            else
+              {
+                // OK, so we have gotten a reply to our answer from
+                // one rank. Let us process it, after double checking
+                // that it is indeed one we were still expecting:
+                const auto target = status.MPI_SOURCE;
+                Assert(outstanding_answers.find(target) !=
+                         outstanding_answers.end(),
+                       ExcMessage(
+                         "We were expecting a message from an MPI rank "
+                         "for which we know that messages are outstanding. But "
+                         "we got one from an unexpected source."));
+
+                const int tag_deliver = Utilities::MPI::internal::Tags::
+                  consensus_algorithm_nbx_process_deliver;
+
+                // Then query the size of the message, allocate enough memory,
+                // receive the data, and process it.
+                int message_size;
+                {
+                  const int ierr =
+                    MPI_Get_count(&status, MPI_BYTE, &message_size);
+                  AssertThrowMPI(ierr);
+                }
+                Assert(message_size % sizeof(T2) == 0, ExcInternalError());
+                std::vector<T2> recv_buffer(message_size / sizeof(T2));
+
+                {
+                  const int ierr = MPI_Recv(recv_buffer.data(),
+                                            recv_buffer.size() * sizeof(T2),
+                                            MPI_BYTE,
+                                            target,
+                                            tag_deliver,
+                                            this->comm,
+                                            MPI_STATUS_IGNORE);
+                  AssertThrowMPI(ierr);
+                }
+
+                this->process.read_answer(target, recv_buffer);
+
+                // Finally, remove this rank from the list of outstanding
+                // targets:
+                outstanding_answers.erase(target);
```

Comment:
Could I ask you to use the set only in debug mode. It is a nice check, but in release mode a counter should enough.

Reply:
I removed it from debug mode as well. I'm not particularly fond of doing different things in debug and release mode where certain variables are only valid in debug mode.